File: FlaneerMediaLib/VideoStreaming/ffmpeg/VideoStreamDecoder.cs
using System.Drawing;
using System.Text;
using FFmpeg.AutoGen;
using FlaneerMediaLib.Logging;

using FF = FFmpeg.AutoGen.ffmpeg;

namespace FlaneerMediaLib.VideoStreaming.ffmpeg
{
    /// <summary>
    /// Decodes video frames
    /// </summary>
    public sealed unsafe class VideoStreamDecoder : IDisposable
    {
        private AVFrame* framePtr;
        private AVFrame* receivedFrame;
        private AVPacket* packetPtr;
        private AVFormatContext* avfCtxPtr;
        
        internal AVCodecContext* CodecContextPtr;

        private int frameCount;
        private bool usingHardwareDecoding => CodecContextPtr->hw_device_ctx != null;
        
        /// <summary>
        /// The height and width of the video frame
        /// </summary>
        public Size SourceSize;
        /// <summary>
        /// The pixel format of the source image for decoding
        /// </summary>
        public AVPixelFormat SourcePixelFormat;

        private readonly Logger logger;

        /// <summary>
        /// ctor
        /// </summary>
        public VideoStreamDecoder()
        {
            logger = Logger.GetLogger(this);
        }
        
        /// <summary>
        /// ctor
        /// </summary>
        public VideoStreamDecoder(AVIOContext* avioCtx)
        {
            logger = Logger.GetLogger(this);

            ServiceRegistry.TryGetService<CommandLineArgumentStore>(out var clArgStore);
            var frameSettings = clArgStore.GetParams(CommandLineArgs.FrameSettings);
            var width =  Int16.Parse(frameSettings[0]);
            var height = Int16.Parse(frameSettings[1]);
            
            Init(avioCtx, width, height);
        }

        /// <summary>
        /// Initializes the decoder
        /// </summary>
        internal void Init(AVIOContext* avioCtx, int width, int height)
        {
            //Allocate an AVFormatContext. avformat_free_context() can be used to free the context and everything allocated by the framework within it.
            var ctx = FF.avformat_alloc_context();
            ctx->pb = avioCtx;

            var arbitrarytext = Encoding.UTF8.GetString(Encoding.Default.GetBytes(""));
            var inFmt = FF.av_find_input_format("h264");
            FF.avformat_open_input(&ctx, arbitrarytext, inFmt, null);

            AVCodec* codec = FF.avcodec_find_decoder(AVCodecID.AV_CODEC_ID_H264);
            //Allocate an AVCodecContext and set its fields to default values. The resulting struct should be freed with avcodec_free_context().
            CodecContextPtr = FF.avcodec_alloc_context3(codec);
            
            CodecContextPtr->width = width;
            CodecContextPtr->height = height;
            CodecContextPtr->pix_fmt = AVPixelFormat.AV_PIX_FMT_YUV420P;

            var hwDec = HwDecodeHelper.GetHWDecoder();
            if (hwDec != AVHWDeviceType.AV_HWDEVICE_TYPE_NONE)
            {
                FF.av_hwdevice_ctx_create(&CodecContextPtr->hw_device_ctx, hwDec, null, null, 0);
            }

            FF.avcodec_open2(CodecContextPtr, codec, null);

            packetPtr = FF.av_packet_alloc();
            framePtr = FF.av_frame_alloc();
            if (usingHardwareDecoding)
                receivedFrame = FF.av_frame_alloc();

            SourceSize = new Size(CodecContextPtr->width, CodecContextPtr->height);
            SourcePixelFormat = usingHardwareDecoding ? HwDecodeHelper.GetHWPixelFormat(hwDec) : CodecContextPtr->pix_fmt;

            avfCtxPtr = ctx;
        }

        /// <summary>
        /// Decode the next frame in the sequence
        /// </summary>
        public AVFrame DecodeNextFrame()
        {
            FF.av_frame_unref(framePtr);
            if(usingHardwareDecoding)
                FF.av_frame_unref(receivedFrame);
            
            int error;
            frameCount++;
            logger.Trace($"Decoding frame {frameCount}");
            do
            {
                try
                {
                    do
                    {
                        FF.av_packet_unref(packetPtr);
                        error = FF.av_read_frame(avfCtxPtr, packetPtr);

                        if (error == FF.AVERROR_EOF)
                        {
                            throw new Exception("error == ffmpeg.AVERROR_EOF");
                        }
                    } while (packetPtr->stream_index != 0);

                    FF.avcodec_send_packet(CodecContextPtr, packetPtr);
                }
                finally
                {
                    FF.av_packet_unref(packetPtr);
                }

                error = FF.avcodec_receive_frame(CodecContextPtr, framePtr);
            } while (error == FF.AVERROR(FF.EAGAIN));

            logger.Trace($"Frame {CodecContextPtr->frame_number}" +
                           $"(type={Convert.ToChar(FF.av_get_picture_type_char(framePtr->pict_type))}," +
                           $" size={framePtr->pkt_size} bytes, format={framePtr->format}) " +
                           $"pts {framePtr->pts} key_frame {framePtr->key_frame} (DTS {framePtr->coded_picture_number})");
            
            if (usingHardwareDecoding)
            {
                FF.av_hwframe_transfer_data(receivedFrame, framePtr, 0);
                return *receivedFrame;
            }

            return *framePtr;
        }

        /// <inheritdoc />
        public void Dispose()
        {
            var pFrame = framePtr;
            FF.av_frame_free(&pFrame);

            var pPacket = packetPtr;
            FF.av_packet_free(&pPacket);

            FF.avcodec_close(CodecContextPtr);
            var pFormatContext = avfCtxPtr;
            FF.avformat_close_input(&pFormatContext);
        }
    }
}

File: FFmpegDecoder/VideoStreamDecoder.cpp
#include "VideoStreamDecoder.h"

#include <iostream>

#define FFMPEGCHECKERR(er, name)({if((er) != 0) {std::cout << (name) << AVErr(er) << "\n"; }})

static AVHWDeviceType GetHWDecoder()
{
    return AV_HWDEVICE_TYPE_NONE;

    AVHWDeviceType HWtype;
    std::vector<AVHWDeviceType> availableHWDecoders;

    auto type = AV_HWDEVICE_TYPE_NONE;
    int it = 0;
    int DXVA2 = 0;

    while ((type = av_hwdevice_iterate_types(type)) != AV_HWDEVICE_TYPE_NONE)
    {
        availableHWDecoders.push_back(type);
        if (type == AV_HWDEVICE_TYPE_DXVA2)
            DXVA2 = it;

        it++;
    }

    if (availableHWDecoders.size() == 0)
        return AV_HWDEVICE_TYPE_NONE;

    //Prefer AV_HWDEVICE_TYPE_DXVA2
    int decoderNumber = DXVA2;
    //else use the first suggested HW decoder
    if (decoderNumber == 0)
        decoderNumber = availableHWDecoders[0];

    return availableHWDecoders.size() < decoderNumber + 1 ? AV_HWDEVICE_TYPE_NONE : availableHWDecoders[decoderNumber];
}

static AVPixelFormat GetHWPixelFormat(AVHWDeviceType hWDevice)
{
    switch (hWDevice)
    {
    case AV_HWDEVICE_TYPE_NONE: return  AV_PIX_FMT_NONE;
    case AV_HWDEVICE_TYPE_VDPAU: return  AV_PIX_FMT_VDPAU;
    case AV_HWDEVICE_TYPE_CUDA: return  AV_PIX_FMT_CUDA;
    case AV_HWDEVICE_TYPE_VAAPI: return  AV_PIX_FMT_VAAPI;
    case AV_HWDEVICE_TYPE_DXVA2: return  AV_PIX_FMT_NV12;
    case AV_HWDEVICE_TYPE_QSV: return  AV_PIX_FMT_QSV;
    case AV_HWDEVICE_TYPE_VIDEOTOOLBOX: return  AV_PIX_FMT_VIDEOTOOLBOX;
    case AV_HWDEVICE_TYPE_D3D11VA: return  AV_PIX_FMT_NV12;
    case AV_HWDEVICE_TYPE_DRM: return  AV_PIX_FMT_DRM_PRIME;
    case AV_HWDEVICE_TYPE_OPENCL: return  AV_PIX_FMT_OPENCL;
    case AV_HWDEVICE_TYPE_MEDIACODEC: return  AV_PIX_FMT_MEDIACODEC;
    default: return AV_PIX_FMT_NONE;
    }
}

VideoStreamDecoder::VideoStreamDecoder(AVIOContext* avioCtx)
{
    //Allocate an AVFormatContext. avformat_free_context() can be used to free the context and everything allocated by the framework within it.
    avfCtxPtr = avformat_alloc_context();
    avfCtxPtr->pb = avioCtx;

    auto inFmt = av_find_input_format("h264");
    const char* arbitraryText = "";
    int error = avformat_open_input(&avfCtxPtr, arbitraryText, inFmt, nullptr);

    if(error != 0)
        std::cout << "avformat_open_input: " << AVErr(error) << "\n";

    AVCodec* codec = avcodec_find_decoder(AV_CODEC_ID_H264);
    //Allocate an AVCodecContext and set its fields to default values. The resulting struct should be freed with avcodec_free_context().
    codecContextPtr = avcodec_alloc_context3(codec);
    //TODO: Set this from config
    codecContextPtr->width = 1920;
    codecContextPtr->height = 1080;
    codecContextPtr->pix_fmt = AV_PIX_FMT_YUV420P;

    auto hwDec = GetHWDecoder();
    if (hwDec != AV_HWDEVICE_TYPE_NONE)
    {
        error = av_hwdevice_ctx_create(&codecContextPtr->hw_device_ctx, hwDec, nullptr, nullptr, 0);
        if (error != 0)
            std::cout << "av_hwdevice_ctx_create" << AVErr(error) << "\n";
    }

    error = avcodec_open2(codecContextPtr, codec, nullptr);
    if (error != 0)
        std::cout << "avcodec_open2" << AVErr(error) << "\n";

    packetPtr = av_packet_alloc();
    framePtr = av_frame_alloc();
    if (usingHardwareDecoding())
        receivedFrame = av_frame_alloc();

    SourceSize = Size{ codecContextPtr->height, codecContextPtr->width};
    SourcePixelFormat = usingHardwareDecoding() ? GetHWPixelFormat(hwDec) : codecContextPtr->pix_fmt;
}

void VideoStreamDecoder::Cleanup()
{
    av_frame_free(&framePtr);    
    av_packet_free(&packetPtr);
    avcodec_close(codecContextPtr);
    avformat_close_input(&avfCtxPtr);
}

AVFrame VideoStreamDecoder::TryDecodeNextFrame()
{
    av_frame_unref(framePtr);
    if (usingHardwareDecoding())
        av_frame_unref(receivedFrame);

    int error;
    frameCount++;
    do
    {
        try
        {
            do
            {
                av_packet_unref(packetPtr);
                error = av_read_frame(avfCtxPtr, packetPtr);

                if (error == AVERROR_EOF)
                {
                    throw std::exception();
                }
                else if (error != 0)
                {
                    std::cout << AVErr(error) << "\n";
                }
            } while (packetPtr->stream_index != 0);

            avcodec_send_packet(codecContextPtr, packetPtr);
        }catch (...){}

        av_packet_unref(packetPtr);

        error = avcodec_receive_frame(codecContextPtr, framePtr);
    } while (error == AVERROR(EAGAIN));

    char path[128];
    sprintf_s(path, "Frame %d (type=%c, size=%d bytes, format=%d) pts %d key_frame %d [DTS %d]",
        codecContextPtr->frame_number,
        av_get_picture_type_char(framePtr->pict_type),
        framePtr->pkt_size,
        framePtr->format,
        framePtr->pts,
        framePtr->key_frame,
        framePtr->coded_picture_number);
    std::cout << path << "\n";

    if (usingHardwareDecoding())
    {
        av_hwframe_transfer_data(receivedFrame, framePtr, 0);
        return *receivedFrame;
    }

    return *framePtr;
}


